### PR TITLE
Consistently error when rest element isn't last in tuple type

### LINKED ISSUE #40235: unbounded type is not strict
**TypeScript Version:**  4.0.2

**Search Terms:** unbounded

**Code**

```ts
type Strings = [string, string];
type Numbers = number[];

type Unbounded = [...Strings, ...Numbers, boolean];
//   ^?

const data: Unbounded = ['a', 'b', false, false]; // <-- No error?
console.log(data);
```

**Expected behavior:**
The above `Unbounded` type should allow only the last element in the tuple as boolean, other elements in range `[2, -2]` if available should be numbers only.

**Actual behavior:**
`[...number[], boolean]` is now handled as `<number | boolean>[]`

**Playground Link:** 
https://www.typescriptlang.org/play?#code/C4TwDgpgBAysBOBLAdgcwM5QLxQNroRVQBooCk0BdAbgChRIoA5AVwFsAjCeTHZdrvFw1a9cNACqyDgHsWyACYQF2PADoNcChlIa1rTt3SlZMgDYQAhshEB6W1EcA9APyiAxjOQEoCy8EsALigpWXklFRxcAHJLaNJojnioADNLM3QIUjSMiBFPb3MINTMZVAAKPwCASmogA

**Related Issues:** 
No.

### PR DESCRIPTION
**Pull request by ahejlsberg**
Fixes #40235.

### CODE REVIEW

**Comment by ahejlsberg** [comment]
@typescript-bot test this

**Comment by ahejlsberg** [comment]
Although unrelated to PR, checking to see if stricter circular recursion check in type inference can fix RWC OOM.

**Comment by ahejlsberg** [comment]
@typescript-bot test this

File: tests/baselines/reference/variadicTuples1.errors.txt [hunk 1]
```diff
@@ -496,4 +499,20 @@ tests/cases/conformance/types/tuple/variadicTuples1.ts(354,26): error TS2322: Ty
...
+!!! error TS1256: A rest element must be last in a tuple type.
+    const data: Unbounded = [false, false];
+    
+    type U1 = [string, ...Numbers, boolean];
+                       ~~~~~~~~~~
+!!! error TS1256: A rest element must be last in a tuple type.
+    type U2 = [...[string, ...Numbers], boolean];
+               ~~~~~~~~~~~~~~~~~~~~~~~
+!!! error TS1256: A rest element must be last in a tuple type.
+    type U3 = [...[string, number], boolean];
```
Comment (royling) [inline tests/baselines/reference/variadicTuples1.errors.txt:517]: Not sure if I get this right, does this mean we still have (or bring back) the constraint that a rest element must be at last in a tuple?
In [4.0 doc](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-4-0.html), I can see the below update:
> The second change is that rest elements can occur anywhere in a tuple - not just at the end!
> Previously, TypeScript would issue an error like the following:
> But with TypeScript 4.0, this restriction is relaxed.

Comment (royling) [reply tests/baselines/reference/variadicTuples1.errors.txt:517]: @ahejlsberg

**Pull request merged**